Ahead of time factory creation command

boesing opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes
| BC Break      | yes
| New Feature   | yes
| RFC           | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

Fixes #171

So just to quickly sum this up:

We do have project configuration, so this is not some fictional use-case.
This configuration contains `service_manager` configuration where some services (usually only depending on interfaces) are registered via `ReflectionBasedAbstractFactory`:

```php
return [
    'service_manager' => [
        'factories' => [
            OurService::class => ReflectionBasedAbstractFactory::class,
        ],
    ],
];
```

Even tho, the `ReflectionBasedAbstractFactory` is an abstract factory, it also extends `FactoryInterface` and thus can be used as shown above. 

- One of the "good" things is, that there is no need for a dedicated factory.
- The "bad" thing is, that the factory consumes runtime to use reflection and to determine the appropriate `__construct` argument values

I've played a little bit in our project and I ended up writing some code (which I now used to contribute here) which allows us to use `ReflectionBasedAbstractFactory` in development while providing a CLI command to generate a opcachable factory while putting it on the filesystem during release artifact generation.


Usage would be as follows:

```plain
Description:
  Creates factories which replace the runtime overhead for `ReflectionBasedAbstractFactory`.

Usage:
  servicemanager:generate-aot-factories [<localConfigFilename>]

Arguments:
  localConfigFilename        Should be a path targeting a filename which will be created so that the config autoloading will pick it up. Using a `.local.php` suffix should verify that the file is overriding existing configuration. [default: "config/autoload/generated-factories.local.php"]
```

One needs also to register a directory in the application config:

**config/autoload/global.php**
```php
return [
    \Laminas\ServiceManager\ConfigProvider::CONFIGURATION_KEY_FACTORY_TARGET_PATH => 'path/to/config/autoloading/directory',
];
```

Since the path registered in that configuration key MUST also be added to `composer.json` `autoload` section using `classmap`, having this as a required configuration key to use the CLI command seemed to be a good solution.

```json
{
     "autoload": {"classmap": ["path/to/factory/output"]}
}
```

So after configuring the application as mentioned above, the CLI command could be executed like:
```bash
$ php vendor/bin/laminas servicemanager:generate-aot-factories config/autoload/aot-factories.local.php
```

This will then scan the projects configuration for **all** container configurations (so everything which looks like a container configuration actually: has to be an array, containing key `factories` being an array, containing values with either `class-string<ReflectionBasedAbstractFactory> `or  instance of `ReflectionBasedAbstractFactory`.

So in the end, we should have a bunch of generated factories located in the `path/to/factory/output` directory (there is a sub-directory per "container") PLUS the `aot-factories.local.php` file located in `config/autoload`. 

In the end, running `composer dump-autoload` should "register" all the factories located in `path/to/factory/output` and the next `vendor/bin/laminas servicemanager:generate-aot-factories` run should tell us, that there are no (more) services registered via `ReflectionBasedAbstractFactory` (as the `aot-factories.local.php` should've overriden the initial application config).

🎉

boesing committed:
feature: introduce ahead-of-time factory compiler

There are projects out there which do want to have "autowiring". Since the service manager is not build with that kind of autowiring compatibility, some developers started to use `ReflectionBasedAbstractFactory` so that they can prevent themselves from writing factories.

Since this project does already provide a CLI command to generate factories, having another CLI command which scans the project configuration for services registered via `ReflectionBasedAbstractFactory` should work.

With this change, both `ReflectionBasedAbstractFactory` and `FactoryCreator`-Command do use the same constructor parameter resolving logic. Due to this, we can safely assume that all services registered with a `ReflectionBasedAbstractFactory` can also be generated during CI.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
Actual commit to review:
https://github.com/laminas/laminas-servicemanager/pull/180/commits/bd6a3626a2a5876dfc7c1c7e39ac4a69be2e4991

froschdesign commented on src/Tool/ConstructorParameterResolverInterface.php:
```diff
@@ -0,0 +1,37 @@
+
+use Psr\Container\ContainerInterface;
+
+interface ConstructorParameterResolverInterface
```
Does this mean that more resolvers will be added in the future? (I am thinking here of forms, for example.)

boesing replied:
Actually not, no. This is just meant to be used with services, IMHO more complex services with scalar argunemt constructors or with a "injection strategy" should stay as is. 

At least I am not planning of supporting more than services available via the servicemanager. 
So anything pluginmanager related wont work with reflection as of now as well. 
But at least having that interface will open up 3rd party to implement more complex resolving strategies. But that wont work with the `ReflectionBasedAbstractFactory` which then requires projects to generate the factories at any time before running the code - thats usually pain in development tho.

froschdesign replied:
> But at least having that interface will open up 3rd party to implement more complex resolving strategies.

👍🏻

bcremer commented:
I like that proposed feature pretty much :+1: 

But I prefer to keep generated files in a dedicated directory (`./data/generated`).
I can archive that with your proposed feature by adding a new `PhpFileProvider`.

```php
$generatedFactoriesConfig = [
    ServiceManagerConfigProvider::CONFIGURATION_KEY_FACTORY_TARGET_PATH => __DIR__ . '/../../data/generated/',
    'factory-override-path' => __DIR__ . '/../data/generated/generated-factories.php'
];

$aggregator = new ConfigAggregator([
    ServiceManagerConfigProvider::class,

    new PhpFileProvider(realpath(__DIR__) . '/autoload/{,*.}{global,local}.php'),

    // Add config for generated factories
    new ArrayProvider($generatedFactoriesConfig),

    // Add generated factories
    new PhpFileProvider($generatedFactoriesConfig['factory-override-path']),
]);

return $aggregator->getMergedConfig();
```

One feature I'd like to propose is adding `localConfigFilename` to the configuration so I don't have to provide it in the `generate-aot-factories` call.

```bash
php bin/console.php servicemanager:generate-aot-factories  data/generated/generated-factories.php
```

How about `ServiceManagerConfigProvider::CONFIGURATION_KEY_FACTORY_LOCAL_CONFIG_FILENAME`?

WDYT?

boesing commented:
@bcremer thats exactly how it is proposed here. You have to configure `ConfigProvider:: CONFIGURATION_KEY_FACTORY_TARGET_PATH` to target to a path where the files have to be stored.
If that is missing, executing the command will output that the config key is missing.

You have to add it as your `composer.json` needs that directory in its `autoload` section.

If we are talking about the path which has to be added by `arg1` via CLI - that won't make sense to make it configurable as it could differ from CLI. Not sure if I'd like to have that configurable as well, especially because if it is configurable, who do you expect that file to "add" to your config then? The whole config-merge has to be done via either `ConfigAggregator` or the `moduleloader` and frankly, having runtime logic which requires configs and merge them on-top of the already merged configs is nothing I'd like to support.

So imho, its up to the projects to specify a path and - in case the path is not automatically picked-up by configuration merge - make the changes to their `ConfigAggregator`/`modulemanager` implementations to ensure the file is properly loaded.
So IMHO, you can do what you provided above while synchronizing that path with your CLI execution as you still will have to provide that path via CLI argument to the command.

bcremer commented:
I withdraw my idea of having a configuration key to configure the path for the override factory mapping (`config/autoload/generated-factories.local.php`).

For most users the the proposed default is the better choice as there is already the default `PhpFileProvider` in place to pick up that file.

I still support the proposed Pull Request from @boesing and would like to see it going forward.

boesing committed:
qa: add native return-types to `ReflectionBasedAbstractFactory` methods

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: mark CLI command as internal

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: add documentation for `Ahead of Time Factories`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented on docs/book/ahead-of-time-factories.md:
```diff
@@ -0,0 +1,105 @@
+
+- Since 4.0.0
+
+In addition to the already existing [Reflection Factory](TODO), one can create factories for those services using `ReflectionBasedAbstractFactory` before deploying the project to production.
```
@froschdesign How can I link from here to the documentation page of `Reflection Factory`?

froschdesign replied:
The same folder, then:

```markdown
[Reflection Factory](reflection-abstract-factory.md)
```

Reference: https://www.mkdocs.org/user-guide/writing-your-docs/#linking-to-pages

boesing committed:
docs: re-order menu items so that there is a `Reflection-based Abstract Factory` category

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: enhance documentation regarding CI pipeline

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

froschdesign commented on docs/book/ahead-of-time-factories.md:
```diff
@@ -0,0 +1,106 @@
+# Ahead of Time Factories 
```
We need versioned documentation, so this must be placed in a subdirectory: `docs/book/v4/ahead-of-time-factories.md`
Create the folder and add the file, I will do the rest.

boesing replied:
Done. I duplicated all files and created both `v3` and `v4` folders.

froschdesign replied:
Great, I will do the rest.

boesing committed:
docs: versionize documentation

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: reference correct file so that mkdocs will generate proper link to that page

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

bcremer commented on docs/book/v4/ahead-of-time-factories.md:
```diff
@@ -0,0 +1,106 @@
+
+When the project is executed having all the files in-place, the generated factory classes are picked up instead of the `ReflectionBasedAbstractFactory` and thus, no additional runtime side-effects based on `Reflection` will occur.
+
+Ensure that both `<target for generated factory config>` file and the directory (including sub-directories and files) configured within `ConfigProvider::CONFIGURATION_KEY_FACTORY_TARGET_PATH`` is being picked up when generating the artifact which is deployed to production. 
```
A superfluous back tick is breaking the formatting here:

```diff
- `ConfigProvider::CONFIGURATION_KEY_FACTORY_TARGET_PATH``
+ `ConfigProvider::CONFIGURATION_KEY_FACTORY_TARGET_PATH`
```

boesing committed:
docs: remove superfluous backtick

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: remove superfluous whitespaces

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: normalize some parts of the migration guide to match markdown rules

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: purge migration guide for v4

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: normalize ahead-of-time-factories.md to match markdown rules

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: remove trailing punctuation

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: mark `AheadOfTimeCompiledFactory#__construct` as internal

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: verify that `localConfigFilename` is actually writable

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: only allow array configuration to be passed

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add unit tests for `AheadOfTimeFactoryCreatorCommand`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: modify config type to `array`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: add dedicated console documentation for the ahead of time factories CLI command

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactor: early return invokable when there are no required parameters

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: apply coding standard to `AheadOfTimeFactoryCompiler`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
refactor: move unit tests from `ReflectionBasedAbstractFactory` to `ConstructorParameterResolver`

The `ReflectionBasedAbstractFactory` should only test itself rather than integration testing the constructor parameter resolver. Tests which verified the appropriate functionality from `ConstructorParameterResolver` were moved to a dedicated test class.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: move more complex tools into dedicated namespaces

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add failing test for already existing factories

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: detect already existing factories

In case there are autoloadable factories with the same class-name as the generated  factory, the factory generation should be aborted to avoid autolaoding conflicts.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: do not allow factory compilation for enums

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add tests for the `AheadOfTimeFactoryCompiler`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add some more unit tests for `AheadOfTimeFactoryCompiler`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add enum issues to psalm baseline

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: fix line-length coding-standard issue

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: remove superfluous whitespace at the end of the line

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

froschdesign reviewed (approved):
The documentation is good, I will take over all further changes, because it would go beyond the scope here.
